lib/galaxy/model/migrations/README.md

```diff
@@ -0,0 +1,70 @@
+
+Since Galaxy uses branch labels to distinguish between the galaxy and the install models, in most cases, you'll need to identify the target branch to which your command should be applied. Use branch labels: `gxy` for the galaxy model, and `tsi` for install model.
+
+### Examples of usage:
+
+Remember to first backup your database(s).
+
+#### To upgrade:
+```
+./migrate_db.sh upgrade gxy@head  # upgrade gxy to head revision
+./migrate_db.sh upgrade gxy@+1  # upgrade gxy to 1 revision above current
+./migrate_db.sh upgrade [revision identifier]  # upgrade branch to a specific revision
+./migrate_db.sh upgrade [revision identifier]+1  # upgrade branch to 1 revision above specific revision
+./migrate_db.sh upgrade heads  # upgrade gxy and tsi to head revisions
+```
+
+#### To downgrade:
+```
+./migrate_db.sh downgrade gxy@base  # downgrade gxy to base (empty db with empty alembic table)
+./migrate_db.sh downgrade gxy@-1  # downgrade gxy to 1 revision below current
+./migrate_db.sh downgrade [revision identifier]  # downgrade branch to a specific revision
+./migrate_db.sh downgrade [revision identifier]-1  # downgrade branch to 1 revision below specific revision
+```
+Check [Alembic documentation](https://alembic.sqlalchemy.org/en/latest/branches.html#working-with-branches) for more examples.
+
+Note: relative upgrades and downgrades without a revision identifier are not supported - i.e., you cannot `upgrade +1` or `downgrade -1` without providing a revision identifier. However, you can upgrade both branches to their latest versions (head revisions) without providing a branch label: `upgrade heads`.
+
+### Upgrading from SQLAlchemy Migrate
+
+Galaxy no longer supports SQLAlchemy Migrate. To upgrade to Alembic, follow these steps:
+
+1. Backup your database(s).
+
+2. Verify that your database is at the latest SQLAlchemy Migrate version. If you have a combined database, it should be version 180 (check the `migrate_version` table). If you have separate galaxy model and install model databases, your galaxy version should be 180, and your install model version should be 17. 
+
+3. If your database is not current, you need to update it first. For that, you'll need to check out a previous version of the code base: use commit 7328ca0da9977bc31cd074464a2e59411a5cdcf8 or later, but prior to [TODO: insert Alembic commit]. Then you can run the old `manage_db.sh` script to bring your database(s) up to date. Once your database has the latest SQLAlchemy Migrate version, you can check out the version of the code base that you had prior to downgrading, and proceed to migrating to Alembic.
+
+4. If you want Alembic to upgrade your database automatically, you can set the [`database_auto_migrate`](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/config_schema.yml#L170) configuration option and simply start Galaxy. Otherwise, use the `migrate_db.sh` script.
```

Comment:
Does that work if you're not on the latest migration ?

Reply:
It won't work if you are not on the latest SQLAlchemy Migrate migration. For that to work, the system would have to first use SQLAlchemy Migrate to upgrate to *that* latest version, and after that pass the control onto Alembic. However, SQlAlchemy Migrate won't run under SA 2.0 (it would break once we enable the future flags in our SA 1.4 >> 2.0 migration process). 
If you look at the whiteboard diagram above, the check for `database_auto_migrate` happens only in one place (bottom left corner) - and at that step, the database is either under alembic version control (in some form - i.e., it varies), or it has the latest SQLAlchemy Migrate version. So we ask to *first* ensure the database is current for the previous release, which, I hope, is a reasonable prerequisite.

Reply:
Hmm .... so we know there's a lot of "dependency-rot" happening over the years, and this does mean that ideally 21.09 remains installable for ~ 5-10 years ... which given past performance seems unlikely. Maybe we should create a docker image that can upgrade Galaxy to these points, for someone needing to make the jump in the future.

Reply:
I'm probably not picturing this correctly.. If we have an instance < 22.01, that instance still runs on the old sqlalchemy migrate code, and sqlalchemy 1.4. So the upgrade path would be the same:
1. `./manage_db.sh upgrade --version=179` (migration handled by sqlalchemy migrate; code base is on SA 1.4)
2. check out new release
3. `./migrate_db.sh upgrade heads` (migration handled by alembic; code base is on SA 1.4 or 2.0)

So, if someone needs to jump from an older (previously supported)  version to 22.01 or beyond, I don't see a problem: they won't be in a situation when the db is under sqlalchemy migrate while their code base is using sqlalchemy 2.0. Or am I missing a use case?

Reply:
Some instances get an update every 2 years (or even less often). So 21.09 (the last one that can migrate to 179) needs to remain installable. That could be a challenge. IOW try installing 17.09, I believe you'll have a hard time doing that.

Reply:
***Very good to know!*** I'm trying this now.

Reply:
OK, this should be fixed. Works as expected: start with 18.01, then `./manage_db.sh upgrade`, then checkout 21.09 (rebuilding .venv), then `./manage_db.sh upgrade`, then checkout alembic branch (aka 22.01), and again `./manage_db.sh upgrade`. 

Along the way discovered a regression: upgrading from dev raised an error when executing the `vault` table revision script (in dev that table was added in db version 180, but not in 21.09). So that's fixed in 6ffed8f.

And a regression on this regression, if we were to:
step 1: upgrade from dev to 22.01 (vault table not created since it's already there)
step 2: downgrade from 22.01 back to dev (vault table dropped)
step 3: in dev, downgrade to a previous db version > error! (trying to drop nonexistent table)

Because after step 2, we'd end up with the current SQLAlchemy Migrate db version 180, which assumes the existence of the `vault` table (which was dropped in step 2 by Alembic!). I'll fix this on dev in a separate PR.